src/core/index.ts

```diff
@@ -339,28 +339,30 @@ export class ZarrArray<StoreGetOptions = any> {
     // create promise queue with concurrency control
     const queue = new PQueue({ concurrency: concurrencyLimit });
 
+    const allTasks = [];
+
     if (progressCallback) {
 
       let progress = 0;
       let queueSize = 0;
       for (const _ of indexer.iter()) queueSize += 1;
       progressCallback({ progress: 0, queueSize: queueSize });
       for (const proj of indexer.iter()) {
-        (async () => {
+        allTasks.push((async () => {
           await queue.add(() => this.chunkGetItem(proj.chunkCoords, proj.chunkSelection, out, proj.outSelection, indexer.dropAxes, storeOptions));
           progress += 1;
           progressCallback({ progress: progress, queueSize: queueSize });
-        })();
+        })());
       }
```

Comment:
Hm, could we just use the queue but remove the async IIFE?

```js
queue.add(async () => {
  await this.chunkGetItem(proj.chunkCoords, proj.chunkSelection, out, proj.outSelection, indexer.dropAxes, storeOptions);
  progress += 1;
  progressCallback({ progress: progress, queueSize: queueSize });
});
```

Reply:
And then keep `queue.onIdle`?

Reply:
I tried to do this and follow the pattern in the rest of the changes, but it brought back the same problem in my use case (trying to cancel chunk requests).  So maybe it's something about how PQueue.onIdle works...?

Reply:
I wonder if the last part of this discussion https://github.com/sindresorhus/p-queue/issues/26 can inform the use of onIdle here.

Reply:
I've looked into the implementation of p-queue while implementing this PR and it seems like the only way to correctly throw on the task error is to await the return of the `queue.add` method, [see here](https://github.com/sindresorhus/p-queue/blob/ce0c36440a28187ff3b3c0e0c91ab6d0d8093b05/source/index.ts#L281)
awaiting the `queue.onIdle` won't throw on error. So it seemed to me you have to create an Array of queue.add return promises and use await Promise.all on this Array.

Reply:
Ok! Thank you for looking into it and for your patience in my response. Would you mind still wrapping the full task in the `queue.add`:

```javascript
  const task = queue.add(async () => {
    await this.chunkGetItem(proj.chunkCoords, proj.chunkSelection, out, proj.outSelection, indexer.dropAxes, storeOptions);
    progress += 1;
    progressCallback({ progress: progress, queueSize: queueSize });
  });
  tasks.push(task);


// ... 

await promise.all(tasks);
```

Reply:
Thanks for the suggestion, I've update the code to remove the iife and move the progress callback inside the queued task.

I didn't await for the progressCallback though since it is typed as returning void, I suggest this is done in a different PR?

Reply:
Great. Thank you!